File: netDxf/IO/DxfWriter.StoredEnvelopes.cs
using System.IO;
using System.Linq;
using netDxf.Objects;
using netDxf.Collections;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private bool WriteStoredEnvelopePayload(DxfDatabaseObject item)
        {
            if (item is DxfSpatialIndex index)
            {
                this.chunk.Write(100, "AcDbIndex"); this.chunk.Write(40, index.Timestamp); this.chunk.Write(100, "AcDbSpatialIndex");
            }
            else if (item is DxfVbaProject project)
            {
                this.chunk.Write(100, "AcDbVbaProject"); this.chunk.Write(90, project.DataLength);
                foreach (byte[] chunk in project.StoredChunks) this.chunk.Write(310, chunk);
            }
            else return false;
            return true;
        }
        private void PrepareStoredEnvelopeClasses(DxfClassCollection definitions)
        {
            if (!this.doc.Objects.Items.Any(item => item is DxfSpatialIndex)) return;
            const string name = "SPATIAL_INDEX";
            int count = this.doc.Objects.Items.Count(item => item.CodeName == name);
            if (definitions.Contains(name))
            {
                DxfClass definition = definitions[name];
                if (definition.CppClassName != "AcDbSpatialIndex" || definition.IsEntity)
                    throw new InvalidDataException("CLASS conflicts with typed SPATIAL_INDEX.");
                definition.InstanceCount = count;
            }
            else definitions.Add(new DxfClass(name, "AcDbSpatialIndex", "ObjectDBX Classes") { ProxyFlags = 0, IsEntity = false, InstanceCount = count });
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.StoredEnvelopes.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfSpatialIndex } from '../Objects/DxfSpatialIndex.js';
import { DxfVbaProject } from '../Objects/DxfVbaProject.js';
import { DxfClass } from '../DxfClass.js';
import { InvalidDataException } from '../../runtime/Errors.js';
export function WriteStoredEnvelopePayload(chunk,version,item){
  if(item instanceof DxfSpatialIndex){chunk.Write(100,'AcDbIndex');chunk.Write(40,item.Timestamp);chunk.Write(100,'AcDbSpatialIndex');}
  else if(item instanceof DxfVbaProject){chunk.Write(100,'AcDbVbaProject');chunk.Write(90,item.DataLength);for(const data of item.StoredChunks)chunk.Write(310,data);}
  else return false;return true;
}
export function PrepareStoredEnvelopeClasses(document,definitions){
  const count=Array.from(document.Objects.Items).filter(i=>i.CodeName==='SPATIAL_INDEX').length;
  if(!Array.from(document.Objects.Items).some(i=>i instanceof DxfSpatialIndex)){
    if(definitions.Contains('SPATIAL_INDEX')&&definitions.get_Item('SPATIAL_INDEX').CppClassName==='AcDbSpatialIndex'&&!definitions.get_Item('SPATIAL_INDEX').IsEntity)definitions.get_Item('SPATIAL_INDEX').InstanceCount=count;return;
  }
  if(definitions.Contains('SPATIAL_INDEX')){const definition=definitions.get_Item('SPATIAL_INDEX');
    if(definition.CppClassName!=='AcDbSpatialIndex'||definition.IsEntity)throw new InvalidDataException('CLASS conflicts with typed SPATIAL_INDEX.');definition.InstanceCount=count;
  }else{const definition=new DxfClass('SPATIAL_INDEX','AcDbSpatialIndex','ObjectDBX Classes');definition.ProxyFlags=0;definition.IsEntity=false;definition.InstanceCount=count;definitions.Add(definition);}
}
